Source/DafnyRuntime/DafnyRuntimeJava/src/main/java/dafny/Tuple2.java

```diff
@@ -0,0 +1,53 @@
+package dafny;
+
+public class Tuple2<T0, T1> {
+    private T0 _0;
+    private T1 _1;
+
+    public static String defaultInstanceName = Tuple2.class.toString();
+
+    public Tuple2(){
+    }
+
+    public Tuple2(T0 _0, T1 _1) {
+        this._0 = _0;
+        this._1 = _1;
+    }
+
+    @Override
+    @SuppressWarnings("UNCHECKED_CAST")
+    public boolean equals(Object obj) {
+        if (this == obj) return true;
+        if (obj == null) return false;
+        if (getClass() != obj.getClass()) return false;
+        Tuple2 o = (Tuple2) obj;
+        return this._0.equals(o._0) && this._1.equals(o._1);
+    }
+
+    @Override
+    public String toString() {
+        StringBuilder sb = new StringBuilder();
+        sb.append("(");
+        sb.append(_0 == null ? "null" : _0.toString());
+        sb.append(", ");
+        sb.append(_1 == null ? "null" : _1.toString());
+        sb.append(")");
+        return sb.toString();
+    }
+
+    @Override
+    public int hashCode() {
+        // GetHashCode method (Uses the djb2 algorithm)
+        // https://stackoverflow.com/questions/1579721/why-are-5381-and-33-so-important-in-the-djb2-algorithm
+        long hash = 5381;
+        hash = ((hash << 5) + hash) + 0;
+        hash = ((hash << 5) + hash) + ((long) this._0.hashCode());
```

Comment:
The code in `toString()` is written to anticipate `_0` and `_1` being `null`, but the code in `hashCode()` and in `equals` will crash in such circumstances. If you expect `null`, then change `hashCode()` and `equals()`. If you don't expect `null`, change `toString()`. In either case, document this somewhere in the class.

The same comment goes for Tuple3.java.